tensorflow/python/keras/layers/wrappers.py

```diff
@@ -318,9 +314,7 @@ def compute_mask(self, inputs, mask=None):
     if inner_mask is not None:
       inner_mask_shape = self._get_shape_tuple((-1,), mask, 2)
       inner_mask = K.reshape(inner_mask, inner_mask_shape)
-    input_uid = generic_utils.object_list_uid(inputs)
-    inner_inputs = self._input_map.get(input_uid, inputs)
-    output_mask = self.layer.compute_mask(inner_inputs, inner_mask)
+    output_mask = self.layer.compute_mask(inputs, inner_mask)
```

Comment:
The input might have a different shape since it was reshaped in the call().

```
inner_input_shape = self._get_shape_tuple((-1,), inputs, 2)
# Shape: (num_samples * timesteps, ...). And track the	        
inputs = array_ops.reshape(inputs, inner_input_shape)
```

Reply:
I see the problem there.
Memory usage would be greatly reduced by storing shape instead of the actual inputs and just reshaping again.  That doesn't actually fix the problem though.  

Does compute_mask get called always once, and only once?  It could just remove it from the dict if so, but I can't find anything specific on that.

Reply:
compute_mask is a public API, so it can be called in all different situation, eg two masks can be created if the same layer is called with two different inputs.

Reply:
All the reshape appears to be the same every time, and I don't see any dependencies on anything but the variable "input".

Which means we could perform the same reshape in compute mask. Did I miss anything?

Reply:
The existing code was trying to get the inner_inputs if it is there, which might be reshaped compare to original inputs.

If we want to want cache the inputs, we somehow need to reshape the inputs again, and make sure the layer.compute_mask() is called with same input tensor when layer.call() is invoked.

Reply:
Hmm. I think compute_mask() was not working quite right already.  If you'd already called call(), it would have a cached value and use the reshaped input.  If you hadn't, it would just use the original input.

I'm not sure which is the correct behavior.  I'm would think the reshaped input is correct.

Reply:
Sorry for the long wait. Thinking about this more, I think its actually better to remove the unbounded caching here with a small dip for the performance.